### ISSUE TITLE
linear_transpose involving fft seemingly incorrect

### ISSUE THREAD

**Issue by EelcoHoogendoorn** [google/jax#6223]
I was planning to implement functionality that requires me to be able to take the transpose of a complicated linear function, which includes fourier transforms. I noticed that the fft module seems to implement the rules for transposition; the last few lines of https://jax.readthedocs.io/en/latest/_modules/jax/_src/lax/fft.html seem to pertain to it.

I am not familiar enough with the jax API at present to point out any bugs; but below is a simple test that demonstrates what I believe to be a bug. The build_matrix helper function explicitly constructs the coefficients of the linear function by feeding it all delta functions in sequence. If I linear_transpose my function, that should be identical to the matrix transpose of the built matrix. Yet it doesnt. It seems as if im getting my result in reverse order (plus another off-by-one index bug I think).

Seems to me like a bug in the implementation of the transpose rules for ffts; but again not qualified myself to spot it.

While on the topic, slightly related question: when viewed as linear operators, convolution and correlation are transposed operators. Should I trust jax to figure out efficient transformations along these lines (assuming the underlying fft rules are bug free); or is it likely optimal for me to figure out how to override the linear transpose of a convolution with my own handcrafted correlation functions (and vice versa)?

Code to reproduce:

```python
import numpy as np

import jax
from jax import numpy as jnp
import matplotlib.pyplot as plt


np.random.seed(0)

signal = np.cumsum(np.random.randn(2**8))
signal_jax = jnp.array(signal)

x = np.linspace(-1, 1, len(signal))
psf = np.clip(0.2 - np.abs(x), 0, 1) * (x > 0)
psf /= psf.sum()
psf_jax = jnp.array(psf)

jrfft = jax.jit(jnp.fft.rfft)
jirfft = jax.jit(jnp.fft.irfft)


@jax.jit
def convolve(a, b):
	fa = jrfft(a)
	fb = jrfft(b)
	return jirfft(fa * fb)


@jax.jit
def correlate(a, b):
	"""NOTE: can this be implemented as a transposition rule according to:
	https://jax.readthedocs.io/en/latest/_modules/jax/_src/lax/fft.html
	"""
	fa = jrfft(a).conj()
	fb = jrfft(b)
	return jirfft(fa * fb)


def psf_convolve(psf):
	"""statically bind psf arg"""
	psf = jax.numpy.fft.ifftshift(psf)
	return lambda a: convolve(psf, a)


def psf_correlate(psf):
	"""statically bind psf arg. psf assumed to be centered"""
	psf = jax.numpy.fft.ifftshift(psf)
	return lambda a: correlate(psf, a)


import types

def build_matrix(func, shape):
	"""explicitly evaluate coeeficient matrix of linear operator func by calling it repeatedly with delta functions"""
	i, j = shape
	Z = []
	I = jnp.eye(i, j)
	for r in range(i):
		z = func(I[r])
		Z.append(z)
	return jnp.array(Z)


func = psf_convolve(psf_jax)

arr = types.SimpleNamespace(shape=signal_jax.shape, dtype=np.float32)
func_trans = lambda a: jax.linear_transpose(func, arr)(a)[0]

N = len(signal)

plt.figure()
M = build_matrix(func, (N, N)).T
plt.imshow(M)

plt.figure()
M = build_matrix(func_trans, (N, N))
plt.imshow(M)
plt.show()
```

- [ ] If applicable, include full error messages/tracebacks.

**Comment by skye** [comment]
@mattjj @hawkinsp @shoyer do any of you have cycles to look into this? (Or cc someone else who may)

**Comment by shoyer** [comment]
I'm not quite sure what's going on yet (you may be correct about a bad transpose rule somewhere), but here's a slightly simplified example that seems to reproduce the same issue:
```python
import numpy as np
import jax
import jax.numpy as jnp

def build_matrix(func, size):
  return jax.vmap(func)(jnp.eye(size, size))

def identity(x):
  # return x  # this works
  return jnp.fft.irfft(jnp.fft.rfft(x))  # this doesn't

identity_trans = lambda a: jax.linear_transpose(identity, a)(a)[0]

N = 4
M_forward = build_matrix(identity, N)
M_backward = build_matrix(identity_trans, N).T

print('Forward matrix:')
print(M_forward)
print('Backward matrix:')
print(M_backward)
```
```
Forward matrix:
[[1. 0. 0. 0.]
 [0. 1. 0. 0.]
 [0. 0. 1. 0.]
 [0. 0. 0. 1.]]
Backward matrix:
[[1. 0. 0. 0.]
 [0. 0. 0. 1.]
 [0. 0. 1. 0.]
 [0. 1. 0. 0.]]
```

**Comment by shoyer** [comment]
A bit more experimentation suggests that the culprit may be the transpose of `irfft`, which appears to be off by a complex conjugate:
```

import numpy as np
import jax
import jax.numpy as jnp

def build_matrix(func, size):
  return jax.vmap(func)(jnp.eye(size, size))

def func(x):
  # return jnp.fft.irfft(jnp.fft.rfft(x))
  # return jnp.fft.fft(x).real
  # return jnp.concatenate([jnp.fft.rfft(x).real, jnp.zeros(1)])
  return jnp.fft.irfft(jnp.concatenate([jnp.zeros(1), x[:2] + 1j*x[2:]]))

func_trans = lambda a: jax.linear_transpose(func, a)(a)[0]

N = 4
M_forward = build_matrix(func, N)
M_backward = build_matrix(func_trans, N).T
print('Forward matrix:')
print(M_forward)
print('Backward matrix:')
print(M_backward)
```
```
Forward matrix:
[[ 0.5   0.   -0.5   0.  ]
 [ 0.25 -0.25  0.25 -0.25]
 [ 0.   -0.5   0.    0.5 ]
 [ 0.    0.    0.    0.  ]]
Backward matrix:
[[ 0.5   0.   -0.5   0.  ]
 [ 0.25 -0.25  0.25 -0.25]
 [ 0.    0.5   0.   -0.5 ]
 [ 0.    0.    0.    0.  ]]
```

This makes sense, given that the transpose rule was copied from TensorFlow, which follows a [different convention for complex gradients](https://github.com/google/jax/issues/4891).

**Comment by shoyer** [comment]
See #6232 for a fix to the gradient rule

To come back to your related question:

> While on the topic, slightly related question: when viewed as linear operators, convolution and correlation are transposed operators. Should I trust jax to figure out efficient transformations along these lines (assuming the underlying fft rules are bug free); or is it likely optimal for me to figure out how to override the linear transpose of a convolution with my own handcrafted correlation functions (and vice versa)?

Generally speaking, our experience has been that gradients from automatic differentiation are about as efficient as those written by hand -- and are of course much less painful to write. In the rare cases where a hand-written gradient is more efficient for a particular operation, the efficiency gap can typically be bridged by defining custom derivative rules for the appropriate operators. 

I don't see any reason why that would be different for transpose rules alone, which in JAX are used as part of calculating gradients.

If you do find such a case, I would be curious to learn about it, and it very likely is something that could be fixed :)

**Comment by EelcoHoogendoorn** [comment]
Thanks for the quick turnaround!

Yeah id much rather rely on the auto-transpose; even if I could in theory optimize a multiplication away I am probably happy to pay that price. I suppose the main worry (after correctness) is that somehow jax will think it needs more/different/slower/less numerically stable fft operations than I think I do; in which case a manual override is quickly going to look appealing. 

But ill get there when ill get there, and ill keep testing for performance and correctness. Will let you know if I find anything significant; and if I don't, thanks again for making an awesome library!

**Comment by EelcoHoogendoorn** [comment]
Related to the above:

```
  # The transpose of RFFT can't be expressed only in terms of irfft. Instead of
  # manually building up larger twiddle matrices (which would increase the
  # asymptotic complexity and is also rather complicated), we rely JAX to
  # transpose a naive RFFT implementation.
```

So yeah sounds like there is some significant compromise there already. I was affraid that this was going to be hard to do optimally full-auto given that the normal preferred implementation relies on a whole bunch of 'insider knowledge' of the underlying compiled fft algos.

Pretty new to jax but it seems like the option I have is to register a jax Primitive that implements a convolution, and registers its transpose as a correlation? 

I dont need to be taking gradients of these, just transposes; but as I understand the fft implementation they are not currently supported on the fft side anyway? (dont see the jvp stuff mentioned in the fft module). 

Bunch of stuff to figure out there as a newbie but looks like a good exercise. This looks like a good [resource]https://jax.readthedocs.io/en/latest/notebooks/How_JAX_primitives_work.html?). But am I reinventing the wheel here by chance?

**Comment by shoyer** [comment]
I think you're probably on the right track with looking into a new primitive to define an efficient transpose rule for FFT based convolutions. That would let you skip the cumbersome transpose rules for `rfft`/`irfft`.

There has been some discussion about implementing a FFT based convolutions in JAX: https://github.com/google/jax/issues/5227. If we do that, that would probably be a good place for such a new primitive to exist inside JAX.

I'll note that the existing `lax.conv_general_dilated` may be a good example of what this should look like (it's just implemented in terms of XLA's `Conv` operator, which is optimized for small kernels and hence does not use FFT).

**Comment by shoyer** [comment]
Actually, rather than writing a custom primitive, this might be a good place to try out @froystig's new `linear_call`: https://github.com/google/jax/pull/5781

**Comment by EelcoHoogendoorn** [comment]
Thanks, it seems like linear_call is just what I need; and indeed nice if I wouldnt have to dig into the more intimidating internals to get there. The PR mentions batching is still somewhat of a WIP, but I guess ill see how that works out in practice.
